src/expr.rs

```diff
@@ -35,7 +77,308 @@ pub enum Expr {
+    type Error = Error;
+    fn try_from(v: Expr) -> Result<HashSet<String>, Error> {
+        match v {
+            Expr::Array(v) => {
+                let mut h = HashSet::new();
+                for el in v {
+                    let _ = h.insert(el.to_text()?);
+                }
+                Ok(h)
+            }
+            _ => Err(Error::ExprToGeom()),
+        }
+    }
+}
+
+fn cmp_op<T: PartialEq + PartialOrd>(left: T, right: T, op: &str) -> Result<Expr, Error> {
+    let out = match op {
+        "=" => Ok(left == right),
+        "<=" => Ok(left <= right),
+        "<" => Ok(left < right),
+        ">=" => Ok(left >= right),
+        ">" => Ok(left > right),
+        "<>" => Ok(left != right),
+        _ => Err(Error::OpNotImplemented("Binary Bool")),
+    };
+    match out {
+        Ok(v) => Ok(Expr::Bool(v)),
+        _ => Err(Error::OperationError()),
+    }
+}
+
+fn arith_op(left: Expr, right: Expr, op: &str) -> Result<Expr, Error> {
+    let left = f64::try_from(left)?;
+    let right = f64::try_from(right)?;
+    let out = match op {
+        "+" => Ok(left + right),
+        "-" => Ok(left - right),
+        "*" => Ok(left * right),
+        "/" => Ok(left / right),
+        "%" => Ok(left % right),
+        "^" => Ok(left.powf(right)),
+        _ => Err(Error::OpNotImplemented("Arith")),
+    };
+    match out {
+        Ok(v) => Ok(Expr::Float(v)),
+        _ => Err(Error::OperationError()),
+    }
+}
+
+fn array_op(left: Expr, right: Expr, op: &str) -> Result<Expr, Error> {
+    let left: HashSet<String> = left.try_into()?;
+    let right: HashSet<String> = right.try_into()?;
+    let out = match op {
+        "a_equals" => Ok(left == right),
+        "a_contains" => Ok(left.is_superset(&right)),
+        "a_containedby" => Ok(left.is_subset(&right)),
+        "a_overlaps" => Ok(!left.is_disjoint(&right)),
+        _ => Err(Error::OpNotImplemented("Arith")),
+    };
+    match out {
+        Ok(v) => Ok(Expr::Bool(v)),
+        _ => Err(Error::OperationError()),
+    }
+}
+
 impl Expr {
+    /// Update this expression with values from the `properties` attribute of a JSON object
+    ///
+    ///  # Examples
+    ///
+    /// ```
+    /// use serde_json::{json, Value};
+    /// use cql2::Expr;
+    /// use std::str::FromStr;
+    ///
+    /// let item = json!({"properties":{"eo:cloud_cover":10, "datetime": "2020-01-01 00:00:00Z", "boolfield": true}});
+    ///
+    /// let fromexpr: Expr = Expr::from_str("boolfield = true").unwrap();
+    /// let reduced = fromexpr.reduce(Some(&item));
+    /// let toexpr: Expr = Expr::from_str("true").unwrap();
+    /// assert_eq!(reduced, toexpr);
+    ///
+    /// let fromexpr: Expr = Expr::from_str("\"eo:cloud_cover\" + 10").unwrap();
+    /// let reduced = fromexpr.reduce(Some(&item));
+    /// let toexpr: Expr = Expr::from_str("20").unwrap();
+    /// assert_eq!(reduced, toexpr);
+    ///
+    /// ```
+    pub fn reduce(&self, j: Option<&Value>) -> Expr {
```

Comment:
I'd like to get this to:

```suggestion
    pub fn reduce(self, j: Option<&Value>) -> Expr {
```

It might be non-trival, but I'm going to hack in a local branch and see what I can do. If it works, it'll help remove a bunch of the `clone`s.